File: 프로그래머스/2/84512. 모음 사전/모음 사전.java
import java.util.*;

class Solution {
    private static final String[] words = {"A", "E", "I", "O", "U"};
    
    private List<String> createDict(String word) {
        List<String> list = new ArrayList<>();
        list.add(word);
        
        if (word.length() == 5) {    
            return list;
        }
        
        for (String w: words) {
            list.addAll(createDict(word + w));
        }
        
        return list;
    }
    
    public int solution(String word) {
        List<String> dict = new ArrayList<>();
        for (String w: words) {
            dict.addAll(createDict(w));
        }
        
        return dict.indexOf(word) + 1;
    }
}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.kt
class Solution {
    val dict = listOf('A', 'E', 'I', "O", 'U', ' ')
    
    fun solution(word: String): Int {
        val resultSet = mutableSetOf<String>()
        for (idx in 0..5) {
            resultSet.addAll(dfs(idx, "", mutableSetOf<String>()))
        }
        val dfsResult = resultSet.toList().sorted()
        return dfsResult.indexOf(word)
        
    }
    
    fun dfs(idx: Int, word: String, acc: MutableSet<String>): Set<String> {
        if (dict[idx] == ' ') {
            acc.add(word)
            return acc
        }
        if (word.length == 5) {
            acc.add(word)
            return acc
        }
        for (next in 0..5) {
            acc.addAll(dfs(next, word + dict[idx], acc.toMutableSet()))
        }
        return acc
    }
}